File: src/main/java/ar/com/flow/bankaccount/transaction/Deposit.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.receipt.CreditPrinter;
import ar.com.flow.money.Money;

public class Deposit {
    public static DepositBuilder to(BankAccount creditAccount) {
        return new DepositBuilder(creditAccount);
    }

    public static class DepositBuilder {
        private final BankAccount creditAccount;

        public DepositBuilder(BankAccount creditAccount) {
            this.creditAccount = creditAccount;
        }

        public Transaction amount(Money amountToDeposit) {
            return Transaction.builder()
                    .amount(amountToDeposit)
                    .step(new Step(creditAccount, Deposit.receipt(creditAccount)))
                    .build();
        }
    }

    public static CreditPrinter receipt(BankAccount account) {
        return new CreditPrinter(account, Action.Deposit);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/Deposit.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.bankaccount.domain.transaction.Builder
import ar.com.flow.bankaccount.domain.transaction.Transaction
import ar.com.flow.bankaccount.domain.transaction.receipt.*
import ar.com.flow.bankaccount.domain.transaction.steps.Step
import ar.com.flow.money.Money

internal object Deposit {
    fun to(creditAccount: BankAccount): DepositBuilder {
        return DepositBuilder(creditAccount)
    }

    internal class DepositBuilder(private val creditAccount: BankAccount) {
        fun amount(amountToDeposit: Money): Transaction {
            return Builder()
                .amount(amountToDeposit)
                .step(Step(creditAccount, DepositCreditPrint(creditAccount), DepositCreditScratch(creditAccount)))
                .build()
        }
    }
}